src/api/client/session.rs

```diff
@@ -0,0 +1,216 @@
+//! Member websocket session definitions and implementations.
+
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use actix::prelude::*;
+use actix_web::ws;
+use actix_web::ws::CloseReason;
+use hashbrown::HashMap;
+use serde_derive::{Deserialize, Serialize};
+
+use crate::{
+    api::control::member::{Id, MemberRepository},
+    log::prelude::*,
+};
+
+/// How long before lack of client message causes a timeout.
+const CLIENT_IDLE_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// Message for close old member session when reconnect [`Web Client`].
+#[derive(Message)]
+struct Close(Option<CloseReason>);
+
+/// Messages to keep [`Web Client`] alive.
+#[derive(Debug, Message, Deserialize, Serialize)]
+pub enum Heartbeat {
+    #[serde(rename = "ping")]
+    Ping(usize),
+    #[serde(rename = "pong")]
+    Pong(usize),
+}
+
+/// State with repositories addresses
+pub struct WsSessionState {
+    pub members_repo: MemberRepository,
+    pub session_repo: WsSessionRepository,
+}
+
+/// Websocket connection is long running connection, it easier
+/// to handle with an actor.
+#[derive(Debug)]
+pub struct WsSessions {
+    member_id: Id,
+    /// Client must send any text message at least once per 10 seconds
+    /// (CLIENT_IDLE_TIMEOUT), otherwise we drop connection.
+    idle_timeout_handler: Option<SpawnHandle>,
+}
+
+impl WsSessions {
+    /// Creates new [`Member`] session with passed-in [`Member`] ID.
+    pub fn new(member_id: Id) -> Self {
+        Self {
+            member_id,
+            idle_timeout_handler: None,
+        }
+    }
+
+    /// Helper method that set idle timeout handler after every message
+    /// from [`Web Client`].
+    fn set_idle_timeout(&mut self, ctx: &mut <Self as Actor>::Context) {
+        if let Some(handler) = self.idle_timeout_handler {
+            ctx.cancel_future(handler);
+        }
+        self.idle_timeout_handler =
+            Some(ctx.run_later(CLIENT_IDLE_TIMEOUT, |_, ctx| {
+                debug!("Client timeout");
+                ctx.notify(Close(Some(ws::CloseCode::Away.into())));
+            }));
+    }
+}
+
+impl Actor for WsSessions {
+    type Context = ws::WebsocketContext<Self, WsSessionState>;
+
+    /// Start the heartbeat process and store session in repository
+    /// on start [`Member`] session.
+    fn started(&mut self, ctx: &mut Self::Context) {
+        self.set_idle_timeout(ctx);
+        let mut session_repo = ctx.state().session_repo.clone();
+        if let Some(old) =
+            session_repo.add_session(self.member_id, ctx.address())
+        {
+            old.do_send(Close(None));
+        }
+    }
+
+    /// Remove [`Member`] session repository after stopped session.
+    fn stopped(&mut self, ctx: &mut Self::Context) {
```

Comment:
@tyranron as an option, can move `remove_session` from `stopped` to handler `Close`. When member restart, send `Close(Some(ws::CloseCode::Restart.into())))` and match it in handler for do not remove session.

Reply:
@Kirguir it still will have race as `.do_send()` doesn't wait completion and executes after session replacement:

1. `WsSession1` starts
2. `WsSession2` starts and it replaces `WsSession1` in repository.
3. `WsSession1` is notified to close with `old.do_send(Close(None))`.
4. `Handler<Close>` fires for `WsSession1` and removes `WsSession2` from repository (as they have same `member_id`).
5. ...oops!

Reply:
@tyranron проблему я понял. 
> `WsSession1` is notified to close with `old.do_send(Close(None))`.

тут отправляем `old.do_send(Close(Some(ws::CloseCode::Restart.into())))`

Удаление сессии из репозитория переносим из `stopped()` в `impl Handle<Close> for WsSession`.

* `WsSession1` starts
* `WsSession2` starts and it replaces `WsSession1` in repository.
* `WsSession1` is notified to close with `old.do_send(Close(Some(ws::CloseCode::Restart.into())))`.
* `Handler<Close>` fires for `WsSession1` and only stopped `WsSession1` but DONT removes `WsSession2` from repository.

Т.е. в хендлере матчим причину закрытия и если это `Restart`, то не удаляем сессию.  

Хм. Но там все равно остается гонка между `reset_idle_timeout` и рестартом: 
* `WsSession1` starts
* `WsSession2` starts and it replaces `WsSession1` in repository.
* `WsSession1` send Close(Away) before is notified to close from `WsSession2` `old.do_send(Close(Some(ws::CloseCode::Restart.into())))` and removes `WsSession2` from repository.

Хитрый маневр с `replace_session` надо заменить на явную проверку, т.е. сначала `get_session`, если есть - 'old.send(Close)` - ждем закрытия и отмены всех `futures` (idle_handler) и только потом `add_session`. 

P.S. мне вообще не очень нравится эта реализация. Она подразумевает такую логику:

- клиент коннектится,  создается сессия и сама себя ложит в репозиторий
- сессия отправляет сообщение `JoinMember(member_id)` в `Room` (условно), который по `member_id` может взять сессию из репозитория.

Можно же отправить `JoinMember(member_id, Addr<WsSession>)` в `Room`, а он пусть сам разбирается, как и кого в репозитории кранить, кого и когда закрывать. `Room` нам нужен и что бы создать и связать `Peer`ы.

Reply:
@Kirguir yup, the problem here is not communication with `Room`, but the restriction of _only 1_ connection from a `Member`, where `Member` actually is not even a single client device (the same user from 2 devices, for example). We should not impose such restrictions at all.

> * сессия отправляет сообщение `JoinMember(member_id)` в `Room` (условно), который по `member_id` может взять сессию из репозитория.

> Можно же отправить `JoinMember(member_id, Addr<WsSession>)` в `Room`, а он пусть сам разбирается, как и кого в репозитории кранить, кого и когда закрывать. `Room` нам нужен и что бы создать и связать `Peer`ы.

Yes, we're breaking layers encapsulation here. `Member` should not know how WS layer is organized. Ideally, it even shouldn't know if WS layer exists at all (dependency inversion goes here), however, at this moment this will complicate things and we'll do such separation later.

WS layer should not operate with `Member`s and its repository. It should only operate with WebSocket connections and pass data to next layer where the actual `Member`'s business logic is implemented. Vice versa is also true.

`Room` also should be able to work with multiple WebSocket connections for a single `Member`, but with the only one WS connection for a single `Peer`.

That's why I propose just to remove "only 1 WS connection for single `Member`" at this moment and implement this logic later in the `Room` layer where detailed `Peer` and `Member` logic will be placed. In that prespective, we don't even need a `WsSessionRepository` as all `WsSession`s will be saved in concrete `Room`s in future. However, we will need some kind of routing collection, which will route `WsSession` by its credentials to a concrete `Addr<Room>`.

Reply:
ack @alexlapa I think this worths voice discussion.

Reply:
@tyranron ух, это ж в корне меняет дело.

Если взять за начало запрос от streaming/api, то там же сразу указывает Room и роль пользователя.

> WS layer should not operate with `Member`s and its repository.

Т.е. в `ws_index` пользователя не проверяем, а только создаем сессию с `credentials` и отправляем в `RoomRouter`, который ищет мембера в `MemberRepository` по кредам, если есть, смотрим роль. Если publish\owner - создаем Room, если это play\member - отправлем его в существующий Room. А адресс рума отдаем назад в сессию, что бы она знала куда отправлять сообщения от WebClient.

Так Room четко знает кто есть кто, и когда подключается member создает 2 Peer и отправляет PeerCreated owner'у и дальше по протоколу. Т.е. сообщения(Command) принимает Room и отдает на вход Peer(ам), который реализует state machine - принимает на вход Command и возвращает новое состояние и Event, который Room отправляет клиенту.

> where `Member` actually is not even a single client device (the same user from 2 devices, for example)

тут возникает конфликт: как отпределить подключается мембер с другого устройства или переподключается? Надо создавать новый пир или заменить в существующем пире сессию?
В OpenViDu member это и есть соединение клиента с конкретного устройства.

api::control - это только реализация, допустим, gRPC протокола, для общения с внешним управляющим API типа streaming/api или его замена?

Reply:
@Kirguir please, summarize here the discussion we've had at Friday.

Reply:
@tyranron @alexlapa в результате дисскусии решили реализовывать такую схему взаимодействия:
- ссылка содержит как `credentials` пользователя, так и идентификатор Room
- Room создаются Control API и запускаются в виде Actor
- каждый Room хранит у себя своих Members, их WsSession и Peer
- Peer реализуют State machine, которой управляет Room. 
- Room умеет обрабатывать команды и на основе состояния Peer генерировать Event 

Последние 2 пункта кажутся мне не совсем удачными:
- Room для управления Peer, надо знать все про Peer, по сути внутреняя логика state machine выносится в Room. После каждого вызова изменения состояния, надо проверить состояние и сгенерировать Event. Будут тонны лишнего кода. Из изученных реализаций state machine видно, что все стараются спрятать всю логику работы внутри. Room должен быть просто шлюзом для команд и событий, и уметь распознавать команды, т.к. некоторые команды нужно ретранслировать обоим пирам.

Нарисовал вот такую схему. Внизу немного кода. Правда тут Command обрабатывает Room, а Event генерирует Peer
![medea signaling big](https://user-images.githubusercontent.com/2745694/53004982-ef5b5b00-343a-11e9-8d3d-474dec7478e4.png)

Reply:
@Kirguir , @alexlapa , @tyranron , Я думаю, что Вам будет сильно легче, если вы не будете оперировать понятиями Room и Member в контексте медиа-сервера. Очевидно room и member - прерогатива некоего Control API. Все станет несколько логичнее, если здесь вместо них речь пойдет о `peer` и `pipeline` (или stream, если не нравится отсыл к kurento), где peer = {ws_connect, type, state}. А pipeline = только 1 publisher(!)+N subscribers. Тогда (уже на другом уровне). Room содержит N pipelines (бизнесс-логикой можно ограничить это,  например только  1) Member сможет иметь N pipelines , в которых хотя-бы один peer его. И пофиг кто он там и какие у него креды.

Reply:
@Kirguir ,

Исправления по схеме:
1. У `Room` нет `owner'`а. Не могу придумать зачем он нужен. В старом проекте к owner'у вязалось создание / удаление комнаты, тут комната будет существовать пока ее не выкосят через Control API.
2. У `Member`'а нет `Role`. Кто куда публикует прокидывается в пирах. `caller` и `responder` это не роли, это идентификаторы Member'ов.

> * ссылка содержит как `credentials` пользователя, так и идентификатор Room
> * Room создаются Control API и запускаются в виде Actor
> * каждый Room хранит у себя своих Members, их WsSession и Peer

:+1: ,

> * Peer реализуют State machine, которой управляет Room.
> * Room умеет обрабатывать команды и на основе состояния Peer генерировать Event

Это будет решаться в рамках `Implement signaling`. Но, если честно, упомянутой Вами проблемы не вижу. 

> вы не будете оперировать понятиями Room и Member в контексте медиа-сервера. Очевидно room и member - прерогатива некоего Control API.

Поэтому и Member и Room находятся в пакете `api::control`. Нужно понимать, что разрабатываемый медиа-сервер будет не только медиа воротить (как `Kurento`), но также решать задачи room-management'а (`openvidu-server`), где, насколько Вы знаете были и`Session` (`Room`) и `Participant` (`Member`), потому что там без них никак. 

> peer = {ws_connect, type, state}

А так и есть, на самом деле :smile: .

`pipeline = только 1 publisher(!)+N subscribers`

Мысль верная, но пока разговор только о p2p, то любой Pipeline будет иметь 1 pub, и 1 sub. `Peer` будет содержать ссылку на другой `Peer`, явно или через `Track`'и. В итоге `Peer` = `Pipeline`.

Reply:
@alexlapa ,

>>>
Мысль верная, но пока разговор только о p2p,
>>>

А нет разницы p2p или клиент-сервер, в первом случае клиент шлет SDP и кандидатов "соседу", а во втором " соседом" является сервер, стоит даже абстрагировать это, пока не придумал как, но это вполне возможно.

>>>
Нужно понимать, что разрабатываемый медиа-сервер будет не только медиа воротить (как `Kurento`), но также решать задачи room management'а (`openvidu-server`)
>>>

вот это как раз, на мой взгляд, и не правильно. Room не единственная и не самая правильная абстракция, реализующая только функционал чата или игры, а что, если это не нужно? Например речь идет о широковещательном эфире, ну типа телевидения? Зачем там room?

Reply:
@flexconstructor ,

> А нет разницы p2p или клиент-сервер, в первом случае клиент шлет SDP и кандидатов "соседу", а во втором " соседом" является сервер, стоит даже абстрагировать это, пока не придумал как, но это вполне возможно.

Ясно что попытаемся построить единую абстракцию для локальных и удаленных пиров, но, мне кажется, что эта задача будет не столь тривиальная как Вам кажется, и в данный момент в это ввязываться не хочется.

> вот это как раз, на мой взгляд, и не правильно. Room не единственная и не самая правильная абстракция, реализующая только функционал чата или игры, а что, если это не нужно? Например речь идет о широковещательном эфире, ну типа телевидения? Зачем там room?

Ни вижу фундаментальных проблем. Может это будет один эндпонт транслирующий на CDN'ку, может это будет обычный Member с PlayЧтоТоТам без авторизации.